can/bus.py

```diff
@@ -152,13 +152,37 @@ def _recv_internal(
         """
         raise NotImplementedError("Trying to read from a write only bus?")
 
+    def send(
+        self, msgs: Union[Sequence[Message], Message], timeout: Optional[float] = None
+    ):
+        """Transmit a message to the CAN bus.
+
+        Override this method to enable the transmit path.
+
+        :param msgs: A message object of sequence of message objects.
+
+        :param timeout:
+            If > 0, wait up to this many seconds for message to be ACK'ed or
+            for transmit queue to be ready depending on driver implementation.
+            If timeout is exceeded, an exception will be raised.
+            Might not be supported by all interfaces.
+            None blocks indefinitely.
+
+        :raises can.CanError:
+            if the message could not be sent
+        """
+        for msg in self._get_messages_sequence(msgs):
+            self._send_internal(msg, timeout)
+
     @abstractmethod
-    def send(self, msg: Message, timeout: Optional[float] = None):
+    def _send_internal(
+        self, msgs: Union[Sequence[Message], Message], timeout: Optional[float] = None
```

Comment:
Shouldn't the method `_send_internal()` take only a single message at a time? This would enable one to override `_send_internal()` for allow sending messages in general and `send()` to speed things up if supported by the driver/hardware/...

Reply:
That will means duplicating some code when the same driver API is use to send one or more messages.

```python
    def send(self, msgs, timeout=None):
        self._send_internal(self._get_messages_sequence(msgs), timeout)

    def _send_internal(self, msg, timeout=None):
        if not ics.validate_hobject(self.dev):
            raise CanError("bus not open")
        messages = tuple(self._msg_to_spy_msg(can_msg) for can_msg in msg)

        try:
            ics.transmit_messages(self.dev, messages)
        except ics.RuntimeError:
            raise ICSApiError(*ics.get_last_api_error(self.dev))
```

VS

```python
    def send(self, msgs, timeout=None):
        if not ics.validate_hobject(self.dev):
            raise CanError("bus not open")
        messages = tuple(self._msg_to_spy_msg(can_msg) for can_msg in msg)

        try:
            ics.transmit_messages(self.dev, messages)
        except ics.RuntimeError:
            raise ICSApiError(*ics.get_last_api_error(self.dev))

    def _send_internal(self, msg, timeout=None):
        if not ics.validate_hobject(self.dev):
            raise CanError("bus not open")

        try:
            ics.transmit_messages(self.dev, self._msg_to_spy_msg(msg))
        except ics.RuntimeError:
            raise ICSApiError(*ics.get_last_api_error(self.dev))
```

Reply:
Well, the second one does not make sense, we would just not implement `_send_internal()` in that case, right?

Yet another approach, which might be a bit cleaner and more versatile:
`send()` receives some `Message` or sequence of them, and transforms them into a sequence, possibly containing just a single element. It is (pretty much?) never overridden. It takes away the (ugly/error prone) handling of sequences discussed somewhere else in this PR.
Then, `_send_internal()` handles only sequences, and in cases where the driver supports multiple messages, it does so directly, and in other cases it simply starts with a loop `for msg in msgs:`.

Reply:
@felixdivo, this is essentially what I did, except the loop is in the abstract class `send`method.

This is to reduce the the changes. A simple rename of `send` --> `_send_internal` in the derived class vs. rename of `send` --> `_send_internal` and adding wrapping code in a for loop.

Reply:
Seems like we all abandoned this at some point last year ...

Giving it another look now your changes seem perfectly reasonable.